mod-source-record-manager-server/src/main/java/org/folio/services/afterprocessing/InstanceProcessingServiceImpl.java

```diff
@@ -45,36 +45,40 @@ public InstanceProcessingServiceImpl(@Autowired JobExecutionSourceChunkDao jobEx
   @Override
   public Future<Void> process(List<Record> records, String sourceChunkId, OkapiConnectionParams params) {
     if (CollectionUtils.isEmpty(records)) {
-      Future.succeededFuture();
+      return Future.succeededFuture();
     }
     Future<Void> future = Future.future();
-    List<Future> createRecordsFuture = new ArrayList<>();
-    List<Pair<Record, Instance>> recordToInstancePairs = mapToInstance(records);
-    RecordProcessingContext processingContext = new RecordProcessingContext(getRecordsType(records));
-    recordToInstancePairs.forEach(recordToInstancePair -> {
-      Record record = recordToInstancePair.getLeft();
-      Instance instance = recordToInstancePair.getRight();
-      Future<Void> postInstanceFuture = postInstance(instance, params).compose(instanceId -> {
-        processingContext.addRecordContext(record.getId(), instanceId);
-        return Future.succeededFuture();
-      });
-      createRecordsFuture.add(postInstanceFuture);
+    List<Future> postInstanceFutureList = new ArrayList<>();
+    RecordToInstanceMapper mapper = RecordToInstanceMapperBuilder.buildMapper(RecordFormat.getByDataType(getRecordsType(records)));
+    SourceStorageClient sourceStorageClient = new SourceStorageClient(params.getOkapiUrl(), params.getTenantId(), params.getToken());
+    for (Record record : records) {
+      if (record.getParsedRecord() != null && record.getParsedRecord().getContent() != null) {
+        Instance instance = mapper.mapRecord(new JsonObject(record.getParsedRecord().getContent().toString()));
+        Future postInstanceFuture = postInstance(instance, params)
+          .compose(instanceId -> {
+            if (Record.RecordType.MARC.equals(record.getRecordType())) {
+              return additionalInstanceFieldsUtil.addInstanceIdToMarcRecord(record, instanceId)
+                .compose(recordWithAdditionalFields -> updateRecord(recordWithAdditionalFields, sourceStorageClient));
+            } else {
+              return Future.succeededFuture();
+            }
+          });
+        postInstanceFutureList.add(postInstanceFuture);
+      }
+    }
+    CompositeFuture.all(postInstanceFutureList).setHandler(result -> {
+      if (result.failed()) {
+        LOGGER.error("Couldn't create Instance in mod-inventory", result.cause());
+        jobExecutionSourceChunkDao.getById(sourceChunkId, params.getTenantId())
+          .compose(optional -> optional
+            .map(sourceChunk -> jobExecutionSourceChunkDao.update(sourceChunk.withState(JobExecutionSourceChunk.State.ERROR), params.getTenantId()))
+            .orElseThrow(() -> new NotFoundException(String.format(
+              "Couldn't update failed jobExecutionSourceChunk status to ERROR, jobExecutionSourceChunk with id %s was not found", sourceChunkId))));
+        future.complete();
+      } else {
+        future.complete();
```

Comment:
complete future one time after the composite clause, and probably it would be nice to have some comment that this is done in order to continue data import process even if instance creation is failed